kv/kvserver: TestLeaderlessWatcherErrorRefreshedOnUnavailabilityTransition failed

kv/kvserver.TestLeaderlessWatcherErrorRefreshedOnUnavailabilityTransition [failed](https://mesolite.cluster.engflow.com/invocations/default/d212a75b-3824-4ed9-8939-2bf28fe3c24e?testReportRun=1&testReportShard=17&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on release-26.1 @ [58dc199eeb4c28f0802afbbc46d41f0e39429e05](https://github.com/cockroachdb/cockroach/commits/58dc199eeb4c28f0802afbbc46d41f0e39429e05):


```
=== RUN   TestLeaderlessWatcherErrorRefreshedOnUnavailabilityTransition
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestLeaderlessWatcherErrorRefreshedOnUnavailabilityTransition3382409654
    test_log_scope.go:82: use -show-logs to present logs inline
    client_replica_test.go:3146: 
        	Error Trace:	pkg/kv/kvserver/client_replica_test.go:3146
        	Error:      	Should be true
        	Test:       	TestLeaderlessWatcherErrorRefreshedOnUnavailabilityTransition
    panic.go:615: -- test log scope end --
test logs left over in: outputs.zip/logTestLeaderlessWatcherErrorRefreshedOnUnavailabilityTransition3382409654
--- FAIL: TestLeaderlessWatcherErrorRefreshedOnUnavailabilityTransition (1.95s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>deadlock=true</code>
 - <code>run=1</code>
 - <code>shard=17</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLeaderlessWatcherErrorRefreshedOnUnavailabilityTransition.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-58676

--- a/pkg/kv/kvserver/client_replica_test.go
+++ b/pkg/kv/kvserver/client_replica_test.go
@@ -3122,6 +3122,12 @@ func TestLeaderlessWatcherErrorRefreshedOnUnavailabilityTransition(t *testing.T)
 				Server: &server.TestingKnobs{
 					WallClock: manual,
 				},
+				Store: &kvserver.StoreTestingKnobs{
+					// Disable raft ticks from refreshing the leaderless watcher
+					// state, allowing the test full control over the leaderless
+					// watcher state.
+					DisableLeaderlessWatcherRefreshOnRaftTick: true,
+				},
 			},
 		},
 	})

--- a/pkg/kv/kvserver/replica_raft.go
+++ b/pkg/kv/kvserver/replica_raft.go
@@ -1478,9 +1478,11 @@ func (r *Replica) tick(
 	postTickStatus := r.mu.internalRaftGroup.BasicStatus()
 
 	// Refresh the unavailability state on the leaderlessWatcher.
-	r.LeaderlessWatcher.refreshUnavailableState(
-		ctx, postTickStatus.Lead, nowPhysicalTime, r.store.cfg.Settings, r.replicaUnavailableErrorRLocked,
-	)
+	if !r.store.TestingKnobs().DisableLeaderlessWatcherRefreshOnRaftTick {
+		r.LeaderlessWatcher.refreshUnavailableState(
+			ctx, postTickStatus.Lead, nowPhysicalTime, r.store.cfg.Settings, r.replicaUnavailableErrorRLocked,
+		)
+	}
 
 	if preTickStatus.RaftState != postTickStatus.RaftState {
 		if postTickStatus.RaftState == raftpb.StatePreCandidate {

--- a/pkg/kv/kvserver/testing_knobs.go
+++ b/pkg/kv/kvserver/testing_knobs.go
@@ -581,6 +581,10 @@ type StoreTestingKnobs struct {
 	// NodeIsLiveCallbackWorkDone, if set, is called after nodeIsLiveCallback
 	// completes its iteration over all replicas on the store.
 	NodeIsLiveCallbackWorkDone func(livenesspb.Liveness)
+
+	// DisableLeaderlessWatcherRefreshOnRaftTick, if set, disables refreshing
+	// the leaderless watcher's unavailable state during raft ticks.
+	DisableLeaderlessWatcherRefreshOnRaftTick bool
 }
 
 // ModuleTestingKnobs is part of the base.ModuleTestingKnobs interface.

